File: hackerrank/problem solving/algorithms/2 - implementation/19 - the hurdle race/the_hurdle_race.py
# https://www.hackerrank.com/challenges/the-hurdle-race/problem?isFullScreen=true

def main():
    _, max_jump = read_nums()
    hurdles = read_nums()
    print(hurdle_race(hurdles, max_jump))


def read_nums():
    return list(map(int, input().split()))


# n: length of hurdles
# T: O(n)
# S: O(1) extra space
def hurdle_race(hurdles, max_jump):
    return max(0, max(hurdles) - max_jump)


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/19 - the hurdle race/the_hurdle_race.ts
// https://www.hackerrank.com/challenges/the-hurdle-race/problem?isFullScreen=true

function main() {
  const maxJump: number = readNums()[1];
  const hurdles: number[] = readNums();
  console.log(hurdleRace(hurdles, maxJump));
}

function readNums(): number[] {
  return readLine().split(' ').map(Number);
}

// n: length of hurdles
// T: O(n)
// S: O(1) extra space
function hurdleRace(hurdles: number[], maxJump: number): number {
  return Math.max(0, Math.max(...hurdles) - maxJump);
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
const inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}
